### ISSUE TITLE
NPE when trying to use `Geocoder.getFromLocation`

### ISSUE THREAD

**Issue by josecoelho** [robolectric/robolectric#3677]
### Description
`Geocoder` is not able to find the LocationService using ServiceManager.findService()

Constructor:
![screen_shot_2018-01-16_at_13_44_09](https://user-images.githubusercontent.com/1987013/34967631-22fec516-fac9-11e7-94d8-a3bba0989579.png)

Executing the code:
![screen_shot_2018-01-16_at_13_44_22](https://user-images.githubusercontent.com/1987013/34967652-413485e8-fac9-11e7-8f4c-cc98346f2b12.png)

### Steps to Reproduce
using targetSdkVersion 26 or 27 create a test calling the method `getFromLocation` from android.localtion.Geocoder

I've created a [fresh project](https://github.com/josecoelho/robolectric_reproduce_geocoder_issue.git) to reproduce this issue.

```java
@RunWith(RobolectricTestRunner.class)
public class GeocoderReproducingNPE {
    @Test
    public void reproduceNPEWithGeocoder() throws IOException {
        Location location = new Location("test");
        location.setLatitude(12.34);
        location.setLongitude(23.45);

        Geocoder geocoder = new Geocoder(RuntimeEnvironment.application.getApplicationContext(), Locale.getDefault());
        List<Address> geocodeResults = geocoder.getFromLocation(location.getLatitude(), location.getLongitude(), 1);

        assertTrue(true);
    }
}
```

### Robolectric & Android Version
```    
   compileSdkVersion 26
    buildToolsVersion '26.0.2'
    defaultConfig {
        minSdkVersion 14
        targetSdkVersion 26
    }
```
I've tested with targetSdkVersion 25 and had the same problem.

Robolectric:
```
    testImplementation "org.robolectric:robolectric:3.6.1"
    testImplementation "org.robolectric:shadows-playservices:3.6.1"
    testImplementation "org.robolectric:shadows-multidex:3.6.1"
    testImplementation "org.robolectric:shadows-supportv4:3.6.1"
```

### Link to a public git repo demonstrating the problem:
https://github.com/josecoelho/robolectric_reproduce_geocoder_issue.git

**Comment by josecoelho** [comment]
I've fixed my tests creating a new `ShadowGeocoder` based on the current one. I will remove it after fixed.

```java
/**
 * Geocoder Shadow created to deal with the problem of ServiceManager.getService(Context.LOCATION_SERVICE) returning null while using Geocoder class
 * Issue created on Robolectric: https://github.com/robolectric/robolectric/issues/3677
 *
 * based on {@link org.robolectric.shadows.ShadowGeocoder}
 *
 * Created by josecoelho on 16/01/18.
 */
@Implements(Geocoder.class)
public class ShadowGeocoder {
    private static boolean isPresent = true;

    /** Returns true by default, or the last value set by {@link #setIsPresent(boolean)}. */
    @Implementation
    public static boolean isPresent() {
        return isPresent;
    }

    /** Statically sets the value to be returned by {@link #isPresent()}. */
    public static void setIsPresent(boolean value) {
        isPresent = value;
    }

    @Resetter
    public static void reset() {
        isPresent = true;
    }

    /**
     * ServiceManager.getService(Context.LOCATION_SERVICE) is returning null and exploding when trying to test android.location.Geocoder using Robolectric
     * To fix this problem, we are mocking a fixed result.
     *
     * This method should work only when maxResults is set to 1.
     * @param latitude
     * @param longitude
     * @param maxResults
     * @return one address
     */
    @Implementation
    public List<Address> getFromLocation(double latitude, double longitude, int maxResults) {
        if(maxResults > 1) {
            throw new RuntimeException("This is a mock and should be used only for maxResults == 1");
        }

        Address address = new Address(Locale.getDefault());
        address.setLatitude(latitude);
        address.setLongitude(longitude);

        ArrayList<Address> addresses = new ArrayList<>(1);
        addresses.add(address);
        return addresses;
    }

}
````

**Comment by josecoelho** [comment]
And here is a nice and elegant `ShadowGeocoder`. Not sure if is that the way that Robolectric is going as we want to use more Android code.

https://github.com/android-ia/platform_external_robolectric/blob/master/src/main/java/com/xtremelabs/robolectric/shadows/ShadowGeocoder.java

**Comment by jongerrish** [comment]
Ah, you found the old Robolectric v1 code. You're right, we do want to use more real Android code, but this calls out to a system service and from digging down, I think simulating the response as you have done is the right thing to do. We'd probably want to provide a non-Android setter on the shadow to allow the test writer to set the results. Any interest in creating a PR? We're going to cut a release, probably this week, if you can provide a PR I am sure we can squeeze it in :-)

**Comment by josecoelho** [comment]
That makes sense. I will work on a new PR for it. Thanks!

**Comment by jongerrish** [comment]
Cool, thanks.

I would just implement the method you need. Don't forget to add tests 😃

Let me know when you're ready for a review and we'll get it into 3.7 😃

**Comment by jongerrish** [comment]
Closing this issue, the fix will make 3.7 coming next week.
